File: Anar/Anar/Core/LinkParser.swift
import Foundation

/// Parses proxy share links (vmess://, vless://, trojan://, ss://, hysteria2://,
/// tuic://, wireguard://) into ProxyProfile values. Lenient by design — real
/// links in the wild vary a lot.
enum LinkParser {

    enum ParseError: LocalizedError {
        case empty
        case unsupported(String)
        case malformed(String)

        var errorDescription: String? {
            switch self {
            case .empty: return "Empty link."
            case .unsupported(let s): return "Unsupported link type: \(s)"
            case .malformed(let s): return "Malformed link (\(s))."
            }
        }
    }

    // MARK: - Entry points

    /// Parses one link. Throws on failure.
    static func parse(_ raw: String) throws -> ProxyProfile {
        let text = raw.trimmingCharacters(in: .whitespacesAndNewlines)
        guard let scheme = text.components(separatedBy: "://").first?.lowercased(),
              text.contains("://") else { throw ParseError.empty }

        switch scheme {
        case "vmess": return try parseVMess(text)
        case "vless": return try parseStandard(text, type: .vless)
        case "trojan": return try parseStandard(text, type: .trojan)
        case "ss": return try parseShadowsocks(text)
        case "hysteria2", "hy2": return try parseStandard(text, type: .hysteria2)
        case "tuic": return try parseStandard(text, type: .tuic)
        case "wireguard", "wg": return try parseWireGuard(text)
        default: throw ParseError.unsupported(scheme)
        }
    }

    /// Parses many links from subscription content: raw newline list or a single
    /// base64 blob containing newline-separated links. Skips lines it can't parse.
    static func parseMany(_ content: String) -> [ProxyProfile] {
        var body = content.trimmingCharacters(in: .whitespacesAndNewlines)
        // A subscription body is often base64-encoded as a whole.
        if !body.contains("://"), let decoded = decodeBase64(body) {
            body = decoded
        }
        var out: [ProxyProfile] = []
        for line in body.split(whereSeparator: \.isNewline) {
            let s = line.trimmingCharacters(in: .whitespaces)
            guard s.contains("://") else { continue }
            if let p = try? parse(s) { out.append(p) }
        }
        return out
    }

    // MARK: - VMess

    private static func parseVMess(_ text: String) throws -> ProxyProfile {
        let payload = String(text.dropFirst("vmess://".count))
        guard let data = decodeBase64(payload)?.data(using: .utf8),
              let obj = try? JSONSerialization.jsonObject(with: data) as? [String: Any]
        else { throw ParseError.malformed("vmess base64/json") }

        func s(_ k: String) -> String { string(obj[k]) }
        func i(_ k: String) -> Int { intValue(obj[k]) ?? 0 }

        var p = ProxyProfile(type: .vmess)
        p.name = s("ps").isEmpty ? s("add") : s("ps")
        p.server = s("add")
        p.port = i("port")
        p.uuid = s("id")
        p.alterId = i("aid")
        p.encryption = s("scy").isEmpty ? "auto" : s("scy")
        p.network = s("net").isEmpty ? "tcp" : s("net")
        p.path = s("path")
        p.hostHeader = s("host")
        let tls = s("tls")
        p.tls = (tls == "tls" || tls == "reality")
        p.sni = s("sni").isEmpty ? p.hostHeader : s("sni")
        p.fingerprint = s("fp")
        if !s("alpn").isEmpty { p.alpn = s("alpn").components(separatedBy: ",").filter { !$0.isEmpty } }
        if p.network == "grpc", !s("path").isEmpty { p.path = s("path") }
        return p
    }

    // MARK: - VLESS / Trojan / Hysteria2 / TUIC (URI form)

    private static func parseStandard(_ text: String, type: ProxyType) throws -> ProxyProfile {
        guard let parts = URIParts(text) else { throw ParseError.malformed("\(type.display) uri") }

        var p = ProxyProfile(type: type)
        p.server = parts.host
        p.port = parts.port
        p.name = parts.fragment.isEmpty ? parts.host : parts.fragment
        let q = parts.query

        switch type {
        case .vless:
            p.uuid = parts.user
            p.flow = q["flow"] ?? ""
            p.encryption = q["encryption"] ?? "none"
        case .trojan:
            p.password = parts.user
        case .hysteria2:
            p.password = parts.user
            p.obfs = q["obfs"] ?? ""
            p.obfsPassword = q["obfs-password"] ?? ""
        case .tuic:
            // tuic://uuid:password@host:port
            let creds = parts.user.components(separatedBy: ":")
            p.uuid = creds.first ?? ""
            p.password = creds.count > 1 ? creds.dropFirst().joined(separator: ":") : ""
            p.congestionControl = q["congestion_control"] ?? q["congestion"] ?? ""
            p.udpRelayMode = q["udp_relay_mode"] ?? ""
        default: break
        }

        // Transport (shared by vless/trojan)
        p.network = q["type"] ?? "tcp"
        if p.network == "grpc" {
            p.path = q["serviceName"] ?? q["servicename"] ?? ""
        } else {
            p.path = q["path"] ?? ""
            p.hostHeader = q["host"] ?? ""
        }

        // TLS / REALITY
        let security = (q["security"] ?? "").lowercased()
        p.tls = (security == "tls" || security == "reality" || type == .hysteria2 || type == .tuic)
        p.reality = (security == "reality")
        p.sni = q["sni"] ?? q["peer"] ?? p.hostHeader
        p.fingerprint = q["fp"] ?? ""
        p.insecure = (q["allowInsecure"] == "1" || q["insecure"] == "1")
        if let alpn = q["alpn"], !alpn.isEmpty {
            p.alpn = alpn.components(separatedBy: ",").filter { !$0.isEmpty }
        }
        if p.reality {
            p.publicKey = q["pbk"] ?? ""
            p.shortId = q["sid"] ?? ""
        }
        return p
    }

    // MARK: - Shadowsocks

    private static func parseShadowsocks(_ text: String) throws -> ProxyProfile {
        var body = String(text.dropFirst("ss://".count))
        var name = ""
        if let hashIdx = body.firstIndex(of: "#") {
            name = percentDecode(String(body[body.index(after: hashIdx)...]))
            body = String(body[..<hashIdx])
        }
        // Strip query (e.g. ?plugin=...) — plugins are not supported in v1.
        if let qIdx = body.firstIndex(of: "?") { body = String(body[..<qIdx]) }

        var p = ProxyProfile(type: .shadowsocks)

        if let atIdx = body.lastIndex(of: "@") {
            // SIP002: base64url(method:password)@host:port
            let userinfo = String(body[..<atIdx])
            let hostport = String(body[body.index(after: atIdx)...])
            let creds = decodeBase64(userinfo) ?? userinfo
            let cp = creds.components(separatedBy: ":")
            p.method = cp.first ?? ""
            p.password = cp.count > 1 ? cp.dropFirst().joined(separator: ":") : ""
            guard let hp = splitHostPort(hostport) else { throw ParseError.malformed("ss host:port") }
            p.server = hp.0; p.port = hp.1
        } else {
            // Legacy: base64(method:password@host:port)
            guard let decoded = decodeBase64(body) else { throw ParseError.malformed("ss base64") }
            guard let atIdx = decoded.lastIndex(of: "@") else { throw ParseError.malformed("ss legacy") }
            let creds = String(decoded[..<atIdx]).components(separatedBy: ":")
            p.method = creds.first ?? ""
            p.password = creds.count > 1 ? creds.dropFirst().joined(separator: ":") : ""
            guard let hp = splitHostPort(String(decoded[decoded.index(after: atIdx)...])) else {
                throw ParseError.malformed("ss legacy host:port")
            }
            p.server = hp.0; p.port = hp.1
        }
        p.name = name.isEmpty ? p.server : name
        return p
    }

    // MARK: - WireGuard (best-effort)

    private static func parseWireGuard(_ text: String) throws -> ProxyProfile {
        guard let parts = URIParts(text) else { throw ParseError.malformed("wireguard uri") }
        var p = ProxyProfile(type: .wireguard)
        p.server = parts.host
        p.port = parts.port
        p.name = parts.fragment.isEmpty ? parts.host : parts.fragment
        p.privateKey = percentDecode(parts.user)
        let q = parts.query
        p.peerPublicKey = q["publickey"] ?? q["public_key"] ?? q["peer_public_key"] ?? ""
        p.preSharedKey = q["presharedkey"] ?? q["pre_shared_key"] ?? ""
        if let addr = q["address"] ?? q["ip"] {
            p.localAddresses = addr.components(separatedBy: ",").map { $0.trimmingCharacters(in: .whitespaces) }
        }
        if let res = q["reserved"] {
            p.reserved = res.components(separatedBy: ",").compactMap { Int($0.trimmingCharacters(in: .whitespaces)) }
        }
        return p
    }

    // MARK: - Helpers

    /// Lightweight URI splitter: scheme://user@host:port?query#fragment
    private struct URIParts {
        var user = ""
        var host = ""
        var port = 443
        var query: [String: String] = [:]
        var fragment = ""

        init?(_ text: String) {
            guard let range = text.range(of: "://") else { return nil }
            var rest = String(text[range.upperBound...])

            if let hashIdx = rest.firstIndex(of: "#") {
                fragment = percentDecode(String(rest[rest.index(after: hashIdx)...]))
                rest = String(rest[..<hashIdx])
            }
            if let qIdx = rest.firstIndex(of: "?") {
                let qs = String(rest[rest.index(after: qIdx)...])
                rest = String(rest[..<qIdx])
                for pair in qs.split(separator: "&") {
                    let kv = pair.split(separator: "=", maxSplits: 1).map(String.init)
                    if kv.count == 2 { query[kv[0].lowercased()] = percentDecode(kv[1]) }
                    else if kv.count == 1 { query[kv[0].lowercased()] = "" }
                }
            }
            if let atIdx = rest.lastIndex(of: "@") {
                user = String(rest[..<atIdx])
                rest = String(rest[rest.index(after: atIdx)...])
            }
            guard let hp = LinkParser.splitHostPort(rest) else { return nil }
            host = hp.0; port = hp.1
        }
    }

    static func splitHostPort(_ s: String) -> (String, Int)? {
        var str = s
        // IPv6 in brackets: [::1]:443
        if str.hasPrefix("["), let close = str.firstIndex(of: "]") {
            let host = String(str[str.index(after: str.startIndex)..<close])
            let after = String(str[str.index(after: close)...])
            let port = after.hasPrefix(":") ? Int(after.dropFirst()) ?? 443 : 443
            return (host, port)
        }
        if let colon = str.lastIndex(of: ":") {
            let host = String(str[..<colon])
            let port = Int(str[str.index(after: colon)...]) ?? 443
            if !host.isEmpty { return (host, port) }
        }
        return str.isEmpty ? nil : (str, 443)
    }

    private static func string(_ any: Any?) -> String {
        if let s = any as? String { return s }
        if let n = any as? NSNumber { return n.stringValue }
        return ""
    }

    private static func intValue(_ any: Any?) -> Int? {
        if let i = any as? Int { return i }
        if let n = any as? NSNumber { return n.intValue }
        if let s = any as? String { return Int(s) }
        return nil
    }

    private static func percentDecode(_ s: String) -> String {
        s.removingPercentEncoding ?? s
    }

    /// Decodes standard or URL-safe base64, with or without padding.
    static func decodeBase64(_ s: String) -> String? {
        var b64 = s.replacingOccurrences(of: "-", with: "+")
                   .replacingOccurrences(of: "_", with: "/")
                   .trimmingCharacters(in: .whitespacesAndNewlines)
        let rem = b64.count % 4
        if rem > 0 { b64.append(String(repeating: "=", count: 4 - rem)) }
        guard let data = Data(base64Encoded: b64) ?? Data(base64Encoded: s) else { return nil }
        return String(data: data, encoding: .utf8)
    }
}

File: AnarAndroid/app/src/main/java/app/anar/android/core/LinkParser.kt
package app.anar.android.core

import android.util.Base64
import org.json.JSONObject
import java.net.URLDecoder

/** Parses proxy share links into ProxyProfile. Ported from the macOS client. */
object LinkParser {

    class ParseException(message: String) : Exception(message)

    fun parse(raw: String): ProxyProfile {
        val text = raw.trim()
        if (!text.contains("://")) throw ParseException("Not a proxy link")
        return when (text.substringBefore("://").lowercase()) {
            "vmess" -> parseVMess(text)
            "vless" -> parseStandard(text, ProxyType.VLESS)
            "trojan" -> parseStandard(text, ProxyType.TROJAN)
            "ss" -> parseShadowsocks(text)
            "hysteria2", "hy2" -> parseStandard(text, ProxyType.HYSTERIA2)
            "tuic" -> parseStandard(text, ProxyType.TUIC)
            "wireguard", "wg" -> parseWireGuard(text)
            else -> throw ParseException("Unsupported link type")
        }
    }

    /** Parses many links from raw or base64 subscription content. */
    fun parseMany(content: String): List<ProxyProfile> {
        var body = content.trim()
        if (!body.contains("://")) {
            decodeBase64(body)?.let { body = it }
        }
        return body.lineSequence()
            .map { it.trim() }
            .filter { it.contains("://") }
            .mapNotNull { runCatching { parse(it) }.getOrNull() }
            .toList()
    }

    // VMess: base64(JSON)
    private fun parseVMess(text: String): ProxyProfile {
        val payload = text.removePrefix("vmess://")
        val json = decodeBase64(payload) ?: throw ParseException("Bad vmess base64")
        val o = JSONObject(json)
        fun s(k: String) = o.opt(k)?.toString() ?: ""
        fun i(k: String) = o.opt(k)?.toString()?.toIntOrNull() ?: 0
        return ProxyProfile(type = ProxyType.VMESS).apply {
            name = s("ps").ifEmpty { s("add") }
            server = s("add")
            port = i("port")
            uuid = s("id")
            alterId = i("aid")
            encryption = s("scy").ifEmpty { "auto" }
            network = s("net").ifEmpty { "tcp" }
            path = s("path")
            hostHeader = s("host")
            val t = s("tls")
            tls = (t == "tls" || t == "reality")
            sni = s("sni").ifEmpty { hostHeader }
            fingerprint = s("fp")
            if (s("alpn").isNotEmpty()) alpn = s("alpn").split(",").filter { it.isNotEmpty() }
        }
    }

    // VLESS / Trojan / Hysteria2 / TUIC URI form
    private fun parseStandard(text: String, type: ProxyType): ProxyProfile {
        val u = Uri(text)
        val p = ProxyProfile(type = type)
        p.server = u.host
        p.port = u.port
        p.name = u.fragment.ifEmpty { u.host }
        val q = u.query

        when (type) {
            ProxyType.VLESS -> {
                p.uuid = u.user
                p.flow = q["flow"] ?: ""
                p.encryption = q["encryption"] ?: "none"
            }
            ProxyType.TROJAN -> p.password = u.user
            ProxyType.HYSTERIA2 -> {
                p.password = u.user
                p.obfs = q["obfs"] ?: ""
                p.obfsPassword = q["obfs-password"] ?: ""
            }
            ProxyType.TUIC -> {
                val creds = u.user.split(":")
                p.uuid = creds.getOrElse(0) { "" }
                p.password = if (creds.size > 1) creds.drop(1).joinToString(":") else ""
                p.congestionControl = q["congestion_control"] ?: q["congestion"] ?: ""
                p.udpRelayMode = q["udp_relay_mode"] ?: ""
            }
            else -> {}
        }

        p.network = q["type"] ?: "tcp"
        if (p.network == "grpc") {
            p.path = q["serviceName"] ?: q["servicename"] ?: ""
        } else {
            p.path = q["path"] ?: ""
            p.hostHeader = q["host"] ?: ""
        }

        val security = (q["security"] ?: "").lowercase()
        p.tls = security == "tls" || security == "reality" ||
                type == ProxyType.HYSTERIA2 || type == ProxyType.TUIC
        p.reality = security == "reality"
        p.sni = q["sni"] ?: q["peer"] ?: p.hostHeader
        p.fingerprint = q["fp"] ?: ""
        p.insecure = q["allowInsecure"] == "1" || q["insecure"] == "1"
        q["alpn"]?.takeIf { it.isNotEmpty() }?.let { p.alpn = it.split(",").filter { s -> s.isNotEmpty() } }
        if (p.reality) {
            p.publicKey = q["pbk"] ?: ""
            p.shortId = q["sid"] ?: ""
        }
        return p
    }

    private fun parseShadowsocks(text: String): ProxyProfile {
        var body = text.removePrefix("ss://")
        var name = ""
        body.indexOf('#').takeIf { it >= 0 }?.let {
            name = urlDecode(body.substring(it + 1))
            body = body.substring(0, it)
        }
        body.indexOf('?').takeIf { it >= 0 }?.let { body = body.substring(0, it) }

        val p = ProxyProfile(type = ProxyType.SHADOWSOCKS)
        val at = body.lastIndexOf('@')
        if (at >= 0) {
            val userinfo = body.substring(0, at)
            val hostport = body.substring(at + 1)
            val creds = (decodeBase64(userinfo) ?: userinfo).split(":")
            p.method = creds.getOrElse(0) { "" }
            p.password = if (creds.size > 1) creds.drop(1).joinToString(":") else ""
            val hp = splitHostPort(hostport) ?: throw ParseException("ss host:port")
            p.server = hp.first; p.port = hp.second
        } else {
            val decoded = decodeBase64(body) ?: throw ParseException("ss base64")
            val a = decoded.lastIndexOf('@')
            if (a < 0) throw ParseException("ss legacy")
            val creds = decoded.substring(0, a).split(":")
            p.method = creds.getOrElse(0) { "" }
            p.password = if (creds.size > 1) creds.drop(1).joinToString(":") else ""
            val hp = splitHostPort(decoded.substring(a + 1)) ?: throw ParseException("ss legacy host:port")
            p.server = hp.first; p.port = hp.second
        }
        p.name = name.ifEmpty { p.server }
        return p
    }

    private fun parseWireGuard(text: String): ProxyProfile {
        val u = Uri(text)
        val p = ProxyProfile(type = ProxyType.WIREGUARD)
        p.server = u.host
        p.port = u.port
        p.name = u.fragment.ifEmpty { u.host }
        p.privateKey = urlDecode(u.user)
        val q = u.query
        p.peerPublicKey = q["publickey"] ?: q["public_key"] ?: q["peer_public_key"] ?: ""
        p.preSharedKey = q["presharedkey"] ?: q["pre_shared_key"] ?: ""
        (q["address"] ?: q["ip"])?.let { p.localAddresses = it.split(",").map { s -> s.trim() } }
        q["reserved"]?.let { p.reserved = it.split(",").mapNotNull { s -> s.trim().toIntOrNull() } }
        return p
    }

    // Lightweight URI: scheme://user@host:port?query#fragment
    private class Uri(text: String) {
        var user = ""
        var host = ""
        var port = 443
        val query = HashMap<String, String>()
        var fragment = ""

        init {
            var rest = text.substringAfter("://")
            rest.indexOf('#').takeIf { it >= 0 }?.let {
                fragment = urlDecode(rest.substring(it + 1))
                rest = rest.substring(0, it)
            }
            rest.indexOf('?').takeIf { it >= 0 }?.let {
                val qs = rest.substring(it + 1)
                rest = rest.substring(0, it)
                for (pair in qs.split("&")) {
                    val kv = pair.split("=", limit = 2)
                    if (kv.size == 2) query[kv[0].lowercase()] = urlDecode(kv[1])
                    else if (kv.size == 1 && kv[0].isNotEmpty()) query[kv[0].lowercase()] = ""
                }
            }
            rest.lastIndexOf('@').takeIf { it >= 0 }?.let {
                user = rest.substring(0, it)
                rest = rest.substring(it + 1)
            }
            val hp = splitHostPort(rest) ?: throw ParseException("bad host:port")
            host = hp.first; port = hp.second
        }
    }

    private fun splitHostPort(s: String): Pair<String, Int>? {
        if (s.startsWith("[")) {
            val close = s.indexOf(']')
            if (close > 0) {
                val host = s.substring(1, close)
                val after = s.substring(close + 1)
                val port = if (after.startsWith(":")) after.drop(1).toIntOrNull() ?: 443 else 443
                return host to port
            }
        }
        val colon = s.lastIndexOf(':')
        if (colon > 0) {
            val host = s.substring(0, colon)
            val port = s.substring(colon + 1).toIntOrNull() ?: 443
            if (host.isNotEmpty()) return host to port
        }
        return if (s.isEmpty()) null else s to 443
    }

    private fun urlDecode(s: String): String =
        runCatching { URLDecoder.decode(s, "UTF-8") }.getOrDefault(s)

    fun decodeBase64(s: String): String? {
        val cleaned = s.replace("-", "+").replace("_", "/").trim()
        for (flags in intArrayOf(Base64.DEFAULT, Base64.NO_PADDING or Base64.NO_WRAP)) {
            runCatching {
                val bytes = Base64.decode(cleaned, flags)
                return String(bytes, Charsets.UTF_8)
            }
        }
        return null
    }
}
